### PR TITLE
fix(timepicker): fixing tab navigation & handling keyboard arrows

### LINKED ISSUE #1836: Tab-based navigation in timepicker with enabled spinners
### Bug description:

Using a timepicker with spinners enabled, I expect that I can change from the hours field to the minutes field using tab on the keyboard. While this works without spinners, it doesn't work with spinners enabled. Here, pressing tab selects the first spinner control.

### Link to minimally-working plunker that reproduces the issue:

See demo page: https://ng-bootstrap.github.io/#/components/timepicker/examples

Example showing the requested behaviour: http://angular-ui.github.io/bootstrap/#!#%2Ftimepicker

### Version of Angular, ng-bootstrap, and Bootstrap:

Angular: 4.3.6

ng-bootstrap: 1.0.0-beta.2

Bootstrap: 4.0.0-beta

### PR DESCRIPTION
**Pull request by ymeine**
Fixing tab navigation between inputs and handling keyboard arrow up and down to change time.

close #1836

### CODE REVIEW

File: src/timepicker/timepicker.ts [hunk 1]
```diff
@@ -81,51 +86,99 @@ const NGB_TIMEPICKER_VALUE_ACCESSOR = {
...
+            (click)="changeMinute(-minuteStep)"
+            [disabled]="disabled"
+            [class.disabled]="disabled"
+          >
             <span class="chevron bottom"></span>
             <span class="sr-only"  i18n="@@ngb.timepicker.decrement-minutes">Decrement minutes</span>
           </button>
         </div>
-        <div *ngIf="seconds" class="ngb-tp-spacer">:</div>
+        <div *ngIf="seconds" class="ngb-tp-spacer">:</div>=
```
Comment (maxokorokov) [inline src/timepicker/timepicker.ts]: You have added the `=` symbol in the template

screen shot 2018-06-28 at 17 59 51

File: src/timepicker/timepicker.ts [hunk 2]
```diff
@@ -1,11 +1,16 @@
 import {Component, forwardRef, Input, OnChanges, SimpleChanges} from '@angular/core';
 import {ControlValueAccessor, NG_VALUE_ACCESSOR} from '@angular/forms';
 
-import {isNumber, padNumber, toInteger} from '../util/util';
+import {isNumber, padNumber, toInteger, isDefined} from '../util/util';
 import {NgbTime} from './ngb-time';
 import {NgbTimepickerConfig} from './timepicker-config';
 import {NgbTimeAdapter} from './ngb-time-adapter';
 
+enum Key {
```
Comment (maxokorokov) [inline src/timepicker/timepicker.ts]: Could you please rebase, to use common Keys?

**Comment by maxokorokov** [comment]
Hey, @ymeine I just merged #2912 as a simpler implementation for the keyboard support (unfortunately with the same drawbacks as here, but it needs the whole component rewrite I suppose).

What's still missing though is `tabindex` on the arrows. Can I close this or you'll be updating it ?

**Comment by maxokorokov** [comment]
@ymeine, you're right about this doing two more things:

- fixing the cursor that is jumping when pressing up/down arrows
- adding tabindex -1

Though you haven't left any comments about this and I missed the first point while testing your PR.

On the other hand I'd very much prefer adding `(keydown.ArrowUp)="changeHour(hourStep); $event.preventDefault()"` to adding 4 additional methods. 

WDYT about just adding `preventDefault()`s ?

**Comment by maxokorokov** [comment]
@ymeine also this implementation breaks selection with `Shift + ArrowUp/Down`

**Comment by ymeine** [comment]
I will update this PR and simplify the code

**Comment by ymeine** [comment]
I pushed another version. It contains two commits:

- first is a refactoring, since there was a lot of duplicated code!
- second is the actual implementation, with e2e tests

I strongly advise to take the refactoring into account (and am open to suggestions to improve it maybe — naming, etc.) since we clearly have three parts (hour, minute, second) which have the same UI and logic.

**Comment by ymeine** [comment]
So, I finally extracted the refactoring commit to put it in a new PR (see [chore(timepicker): refactoring, common code for hour/minute/second by ymeine · Pull Request #3128 · ng-bootstrap/ng-bootstrap](https://github.com/ng-bootstrap/ng-bootstrap/pull/3128)) since it might spoil the discussion thread here and slow down the adoption of the fix.

Now only the straightforward fix is available for this current PR. Hoping to make it through.

File: e2e-app/src/app/timepicker/navigation/timepicker-navigation.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+describe('Timepicker', () => {
+  let page: Page;
+
+  beforeAll(() => page = new Page());
+  beforeEach(async() => await openUrl('timepicker/navigation'));
+
+  function focusInputBefore() { return page.getInputBefore().click(); }
+
+  function goNext() { return sendKey(Key.TAB); }
```
Comment (maxokorokov) [inline e2e-app/src/app/timepicker/navigation/timepicker-navigation.e2e-spec.ts]: I would argue that these keyboard helpers make it MORE difficult to read the test.

I would prefer to read the test:

```
focus input before
press TAB
expect hourField to be focused
press SHIFT+TAB
...
```

as opposed to using `goNext()` which I have no idea what it does.
What's the point of hiding of what you're actually doing and wrapping one function into another here?

P.S. talking about only `sendKey(...)` here

File: e2e-app/src/app/timepicker/navigation/timepicker-navigation.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+import {Page} from './timepicker-navigation.po';
+
+describe('Timepicker', () => {
+  let page: Page;
+
+  beforeAll(() => page = new Page());
+  beforeEach(async() => await openUrl('timepicker/navigation'));
+
+  function focusInputBefore() { return page.getInputBefore().click(); }
```
Comment (maxokorokov) [inline e2e-app/src/app/timepicker/navigation/timepicker-navigation.e2e-spec.ts]: shouldn't this go to the `Page`?

File: e2e-app/src/app/timepicker/navigation/timepicker-navigation.po.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import {$} from 'protractor';
+
+export type Field = 'hour' | 'minute' | 'second';
+
+export class Page {
```
Comment (maxokorokov) [inline e2e-app/src/app/timepicker/navigation/timepicker-navigation.po.ts]: Could you please rename to `TimepickerPage`, please?
- all other pages have specific names
- having only `Page` makes navigation by class name difficult

**Comment by ymeine** [comment]
Changes done!

**Pull request merged**